### PR TITLE
fix: Improve uncaught error handling

### LINKED ISSUE #243: Potentially catch global `onunhandledrejection` as errors?
Currently Cypress taps into `window.onerror` for failing tests which have script errors.

But we don't tap into `onunhandledrejection` events. I believe we should, but enable this to be configurable.

**API Idea:**

```
Cypress.Errors.listenForUncaughtExceptions(false)
```

```
Cypress.Errors.listenForUnhandledRejections(false)
```

### LINKED ISSUE #8802: Some asynchronously thrown spec errors do not fail test
### Current behavior

Uncaught rejections thrown in tests that use `done` will not fail the test when no `cy` chain is active:

https://github.com/cypress-io/cypress/blob/f65fd61d1e81597cb8b1e3e3d95c73a5ce869dac/packages/server/test/support/fixtures/projects/e2e/cypress/integration/caught_async_sync_test_spec.coffee#L89-L95

Similar behavior exists for other unhandled rejections:

https://github.com/cypress-io/cypress/blob/f65fd61d1e81597cb8b1e3e3d95c73a5ce869dac/packages/server/test/support/fixtures/projects/e2e/cypress/integration/caught_async_sync_test_spec.coffee#L70-L73

Potentially related: some asynchronous test failures are incorrectly attributed to the AUT, not the spec window.

### Desired behavior

Correctly attribute unhandled rejections and exceptions to the spec or the AUT, do not miss handling any errors.

### Test code to reproduce

See #8221

### PR DESCRIPTION
**Pull request by chrisbreiding**
- Closes #243
- Closes #8802
- Partially addresses #8236

### User facing changelog

- Cypress now catches uncaught errors and fails the test even if the app-under-test has defined `window.onerror`.
- Cypress now fails tests if there is an unhandled promise rejection in the app-under-test. Unhandled rejections will trigger the `uncaught:exception` event with the promise as the third argument.

### Additional details

Instead of use `window.onerror` to catch uncaught errors, we now use `window.addEventListener('error')`. Before, if the app overwrote `window.onerror`, we would no longer catch any uncaught errors, but that's no longer the case. 

We also now use `window.addEventListener('unhandledrejection')` to catch and fail if there's an unhandled promise rejection.

### How has the user experience changed?

If a user's app overwrote `window.onerror` or had unhandled rejections, tests will fail that trigger errors/unhandled rejections. They will need to fix those errors in their app or return false from `uncaught:exception` to get those tests to pass.

### PR Tasks

- [x] Have tests been added/updated?
- [x] Has the original issue or this PR been tagged with a release in ZenHub? 
- [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? https://github.com/cypress-io/cypress-documentation/pull/3615
- [x] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- N/A Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

File: packages/runner/injection/index.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
+/**
+ * This is the entry point for the script that gets injected into
+ * the AUT. It gets bundled on its own and injected into the <head>
+ * of the AUT by `packages/proxy`.
+ *
+ * If adding to this bundle, try to keep it light and free of
+ * dependencies.
+ */
```
Comment (flotwig) [inline packages/runner/injection/index.js:8]: i'm happy that this is no longer embedded as a string, good stuff

File: packages/driver/src/cy/commands/xhr.js [hunk 2]
```diff
@@ -214,9 +214,8 @@ const startXhrServer = (cy, state, config) => {
         log.snapshot('error').error(err)
       }
 
-      // re-throw the error since this came from AUT code, and needs to
-      // cause an 'uncaught:exception' event. This error will be caught in
-      // top.onerror with stack as 5th argument.
+      // fail the test
+      // cy.fail(err)
```
Comment (flotwig) [inline packages/driver/src/cy/commands/xhr.js:218]:
```suggestion
```

**Review by flotwig — APPROVED** [review]
Changes make sense to me, code looks good too :+1:

**Comment by zwass** [comment]
Hi @chrisbreiding is there an example of how to disable this?

**Comment by zwass** [comment]
Thank you!

**Pull request merged**